File: bash/fifo/guesser.cs
using System;

public class Guesser
{
    public static void Main(string[] args)
    {
        int left=0;
        int right=101;
        int myGuess;
        string response;
        while (true) {
            myGuess = (right+left)/2;
            Console.WriteLine(myGuess);
            Console.Error.WriteLine($"Guesser - Trying {myGuess}, my bounds being {left}...{right} (excluded)");
            response = Console.ReadLine();
            if (string.IsNullOrEmpty(response)) {
                continue;
            }
            Console.Error.WriteLine($"Guesser - Handling input '{response}'");
            if (response==">") {
                left = myGuess;
                continue;
            }
            if (response=="<") {
                right = myGuess;
                continue;
            }
            if (response=="!") {
                Console.Error.WriteLine($"Guesser - Guessed it! :-)");
                break;
            }
        }
    }
}

File: bash/fifo/guesser.py
#!/usr/bin/env python3

import random
import sys

def printOnStderr(text):
    print(text, file=sys.stderr)

def run():
    left=0;
    right=101;
    while True:
        myGuess = (right+left)//2;
        print(myGuess);
        printOnStderr(f"Guesser - Trying {myGuess}, my bounds being {left}...{right} (excluded)");
        response = input();

        printOnStderr(f"Guesser - Handling input '{response}'");
        if (response==">"):
            left = myGuess;
            continue;
        if (response=="<"):
            right = myGuess;
            continue;
        if (response=="!"):
            printOnStderr(f"Guesser - Guessed it! :-)");
            break;

if __name__=="__main__":
    run()
